doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,223 @@
+   string "user:/a/brightness"
+signal time=1666993066.750161 sender=:1.364 -> destination=(null destination) serial=3 path=/org/libelektra/configuration; interface=org.libelektra; member=KeyAdded
+   string "user:/a/saturation"
+signal time=1666993066.750551 sender=:1.364 -> destination=(null destination) serial=4 path=/org/libelektra/configuration; interface=org.libelektra; member=KeyAdded
+   string "user:/a/test"
+signal time=1666993066.750628 sender=:1.364 -> destination=(null destination) serial=5 path=/org/libelektra/configuration; interface=org.libelektra; member=KeyDeleted
+   string "user:/b/background"
+signal time=1666993066.750866 sender=:1.364 -> destination=(null destination) serial=6 path=/org/libelektra/configuration; interface=org.libelektra; member=KeyDeleted
+   string "user:/b/foreground"
+```
+
+As can be seen, the change tracking within the `dbus` plugin wrongly calculates everything below `user:/a` are new keys that have been added, and everything below `user:/b` was removed.
+
+## Constraints
+
+1. As far as possible, we must check for illegal sequences and raise an error when an illegal sequence occurs
+
+## Assumptions
+
+1.
+2.
+3.
+
+## Considered Alternatives
+
+1. Document which sequences are allowed without raising errors in all illegal cases.
+   Least-effort approach, but at least offers some transparency for developers.
+   Developers utilizing Elektra can still ignore it, and may cause problems in certain setups without knowing.
+
+   (Violates constraint 1)
+
+2. Enforce that the `parentKey` used in `kdbSet` is below the one used in the last `kdbGet`.
+   If this is not the case, `kdbSet` will abort and report an error in `parentKey`.
+   Developers might still wrongly mix the sequences, but they will get an error and have to fix it.
+
+   - Adding a `Key * lastGetParent` to `struct _KDB`
+   - Doing `keyCopy (handle->lastGetParent, parentKey, KEY_CP_NAME)` in `kdbGet`
+   - And checking `keyIsSameOrBelow (handle->lastGetParent, parentKey) == 1` in `kdbSet`
+
+3. Enforce that `ks` used in `kdbSet` is the same as in the last `kdbGet`
+   Similar to (2), but we check for the pointer of the `KeySet` instead of having a copy of the parent key, which saves some memory.
+
+   - Adding a `KeySet * lastGetKs` to `struct _KDB`
+   - Doing `handle->lastGetKs = ks` in `kdbGet`
+   - And checking `lastGetKs == ks` in `kdbSet`
+
+4. Change the API so that a single instance of KDB can only contain a single instance of a `KeySet`.
+   This solution completely eliminates problematic sequences.
+
+   ```c
+   // same as before but uses handle->data instead of the ks argument
+   int kdbGet (KDB * handle, Key * parentKey);
+   // same as before but uses handle->data instead of the ks argument
+   int kdbSet (KDB * handle, Key * parentKey);
+
+   KeySet * kdbData (KDB * handle) {
+       return handle->data;
+   }
+   ```
+
+5. Allow arbitrary sequences and let each plugin deal with it on a case-by-case basis.
+
```

Comment:
This is just option 1, with the additional bit that all sequences are allowed.

The change tracking part is being discussed in #4554. _You_ wanted to split the two decisions...

Also: Technically "arbitrary sequences" is not possible anyway, because we already have a constraint:

Before calling `kdbSet (kdb, ksX, parentX)` there must have been a call to `kdbGet (kdb, ksY, parentY)`, such that
- `parentX` is below `parentY` OR
- `parentX` and `parentY` are stored in the same backend.

That's needed so that the backends are correctly initialized.

So what this option should say is to make the above constraint the only one.

Reply:
I agree, I corrected my suggestion to not contain "arbitrary sequences".

> Before calling kdbSet (kdb, ksX, parentX) there must have been a call to kdbGet (kdb, ksY, parentY), such that

Not necessarily. We could have a kdbSet which internally does a kdbGet. But I don't think it should be done in the low-level API. Instead it should be in a higher-level API, also doing necessary merges on conflicts.

Reply:
I've added this alternative.